File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/fragments/social/QuestDetailFragment.java
package com.habitrpg.android.habitica.ui.fragments.social;

import android.app.AlertDialog;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.facebook.drawee.view.SimpleDraweeView;
import com.habitrpg.android.habitica.R;
import com.habitrpg.android.habitica.components.AppComponent;
import com.habitrpg.android.habitica.data.InventoryRepository;
import com.habitrpg.android.habitica.data.SocialRepository;
import com.habitrpg.android.habitica.data.UserRepository;
import com.habitrpg.android.habitica.helpers.RxErrorHandler;
import com.habitrpg.android.habitica.models.inventory.Quest;
import com.habitrpg.android.habitica.models.inventory.QuestContent;
import com.habitrpg.android.habitica.models.members.Member;
import com.habitrpg.android.habitica.models.social.Group;
import com.habitrpg.android.habitica.modules.AppModule;
import com.habitrpg.android.habitica.ui.fragments.BaseMainFragment;
import com.habitrpg.android.habitica.ui.helpers.DataBindingUtils;
import com.habitrpg.android.habitica.ui.helpers.MarkdownParser;

import java.util.List;

import javax.inject.Inject;
import javax.inject.Named;

import butterknife.BindView;
import butterknife.OnClick;

public class QuestDetailFragment extends BaseMainFragment {

    @Inject
    SocialRepository socialRepository;
    @Inject
    UserRepository userRepository;
    @Inject
    InventoryRepository inventoryRepository;
    @Inject
    @Named(AppModule.NAMED_USER_ID)
    String userId;

    @BindView(R.id.quest_title_view)
    TextView questTitleView;
    @BindView(R.id.quest_scroll_image_view)
    SimpleDraweeView questScrollImageView;
    @BindView(R.id.quest_leader_view)
    TextView questLeaderView;
    @BindView(R.id.description_view)
    TextView questDescriptionView;
    @BindView(R.id.quest_participant_list)
    LinearLayout questParticipantList;
    @BindView(R.id.participants_header)
    TextView participantHeader;
    @BindView(R.id.participants_header_count)
    TextView participantHeaderCount;
    @BindView(R.id.quest_participant_response_wrapper)
    ViewGroup questParticipantResponseWrapper;
    @BindView(R.id.quest_leader_response_wrapper)
    ViewGroup questLeaderResponseWrapper;
    @BindView(R.id.quest_accept_button)
    Button questAcceptButton;
    @BindView(R.id.quest_reject_button)
    Button questRejectButton;
    @BindView(R.id.quest_begin_button)
    Button questBeginButton;
    @BindView(R.id.quest_cancel_button)
    Button questCancelButton;
    @BindView(R.id.quest_abort_button)
    Button questAbortButton;

    public String partyId;
    public String questKey;
    private Group party;
    private Quest quest;
    private String begin_quest_message;


    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        return inflater.inflate(R.layout.fragment_quest_detail, container, false);
    }


    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
    }

    @Override
    public void onResume() {
        super.onResume();
        getCompositeSubscription().add(socialRepository.getGroup(partyId).subscribe(this::updateParty, RxErrorHandler.handleEmptyError()));
        getCompositeSubscription().add(inventoryRepository.getQuestContent(questKey).subscribe(this::updateQuestContent, RxErrorHandler.handleEmptyError()));
    }


    private void updateParty(Group group) {
        if (questTitleView == null || group.quest == null) {
            return;
        }
        party = group;
        quest = group.quest;
        setQuestParticipants(group.quest.participants);
        socialRepository.getMember(quest.leader).first().subscribe(member -> {
            if (getContext() != null && questLeaderView != null && member != null) {
                questLeaderView.setText(getContext().getString(R.string.quest_leader_header, member.getDisplayName()));
            }
        }, RxErrorHandler.handleEmptyError());

        if (questLeaderResponseWrapper != null) {
            if (showParticipatantButtons()) {
                questLeaderResponseWrapper.setVisibility(View.GONE);
                questParticipantResponseWrapper.setVisibility(View.VISIBLE);
            } else if (showLeaderButtons()) {
                questParticipantResponseWrapper.setVisibility(View.GONE);
                questLeaderResponseWrapper.setVisibility(View.VISIBLE);
                if (isQuestActive()) {
                    questBeginButton.setVisibility(View.GONE);
                    questCancelButton.setVisibility(View.GONE);
                    questAbortButton.setVisibility(View.VISIBLE);
                } else {
                    questBeginButton.setVisibility(View.VISIBLE);
                    questCancelButton.setVisibility(View.VISIBLE);
                    questAbortButton.setVisibility(View.GONE);
                }
            } else {
                questLeaderResponseWrapper.setVisibility(View.GONE);
                questParticipantResponseWrapper.setVisibility(View.GONE);
            }
        }
    }

    private boolean showLeaderButtons() {
        return party != null && party.quest != null && userId != null && userId.equals(party.quest.leader);
    }

    private boolean showParticipatantButtons() {
        if (user == null || user.getParty() == null || user.getParty().getQuest() == null) {
            return false;
        }
        return !isQuestActive() && user.getParty().getQuest().RSVPNeeded;
    }

    private boolean isQuestActive() {
        return quest != null && quest.active;
    }


    private void updateQuestContent(QuestContent questContent) {
        if (questTitleView == null || !questContent.isManaged()) {
            return;
        }
        questTitleView.setText(questContent.getText());
        questDescriptionView.setText(MarkdownParser.parseMarkdown(questContent.getNotes()));
        DataBindingUtils.INSTANCE.loadImage(questScrollImageView, "inventory_quest_scroll_"+questContent.getKey());
    }

    private void setQuestParticipants(List<Member> participants) {
        if (questParticipantList == null) {
            return;
        }
        questParticipantList.removeAllViews();
        LayoutInflater inflater = (LayoutInflater) getContext().getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        int participantCount = 0;
        for (Member participant : participants) {
            if (quest.active && (participant.getParticipatesInQuest() == null || !participant.getParticipatesInQuest())) {
                continue;
            }
            View participantView = inflater.inflate(R.layout.quest_participant, questParticipantList, false);
            TextView textView = (TextView) participantView.findViewById(R.id.participant_name);
            textView.setText(participant.getDisplayName());
            TextView statusTextView = (TextView) participantView.findViewById(R.id.status_view);
            if (!quest.active) {
                if (participant.getParticipatesInQuest() == null) {
                    statusTextView.setText(R.string.pending);
                    statusTextView.setTextColor(ContextCompat.getColor(getContext(), R.color.gray_200));
                } else if (participant.getParticipatesInQuest()) {
                    statusTextView.setText(R.string.accepted);
                    statusTextView.setTextColor(ContextCompat.getColor(getContext(), R.color.green_100));
                } else {
                    statusTextView.setText(R.string.declined);
                    statusTextView.setTextColor(ContextCompat.getColor(getContext(), R.color.red_100));
                }
            } else {
                statusTextView.setVisibility(View.GONE);
            }
            questParticipantList.addView(participantView);
            if (quest.active || (participant.getParticipatesInQuest() != null && participant.getParticipatesInQuest())) {
                participantCount += 1;
            }
        }
        if (quest.active) {
            participantHeader.setText(R.string.participants);
            participantHeaderCount.setText(String.valueOf(participantCount));
        } else {
            participantHeader.setText(R.string.invitations);
            participantHeaderCount.setText(participantCount + "/" + quest.participants.size());
            begin_quest_message = getString(R.string.quest_begin_message, participantCount, quest.participants.size());
        }
    }

    @Override
    public void onDestroyView() {
        socialRepository.close();
        userRepository.close();
        inventoryRepository.close();
        super.onDestroyView();
    }

    @OnClick(R.id.quest_accept_button)
    public void onQuestAccept() {
        socialRepository.acceptQuest(user, partyId).subscribe(aVoid -> {}, RxErrorHandler.handleEmptyError());
    }


    @OnClick(R.id.quest_reject_button)
    public void onQuestReject() {
        socialRepository.rejectQuest(user, partyId).subscribe(aVoid -> {}, RxErrorHandler.handleEmptyError());
    }

    @OnClick(R.id.quest_begin_button)
    public void onQuestBegin() {
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity())
                .setMessage(begin_quest_message)
                .setPositiveButton(R.string.yes, (dialog, which) -> socialRepository.forceStartQuest(party)
                        .subscribe(quest -> {}, RxErrorHandler.handleEmptyError()))
                .setNegativeButton(R.string.no, (dialog, which) -> {});
        builder.show();
    }

    @OnClick(R.id.quest_cancel_button)
    public void onQuestCancel() {
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity())
                .setMessage(R.string.quest_cancel_message)
                .setPositiveButton(R.string.yes, (dialog, which) -> socialRepository.cancelQuest(partyId)
                        .subscribe(aVoid -> getActivity().getFragmentManager().popBackStack(), RxErrorHandler.handleEmptyError())).setNegativeButton(R.string.no, (dialog, which) -> {});
        builder.show();
    }

    @OnClick(R.id.quest_abort_button)
    public void onQuestAbort() {
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity())
                .setMessage(R.string.quest_abort_message)
                .setPositiveButton(R.string.yes, (dialog, which) -> socialRepository.abortQuest(partyId)
                        .subscribe(aVoid -> getActivity().getFragmentManager().popBackStack(), RxErrorHandler.handleEmptyError())).setNegativeButton(R.string.no, (dialog, which) -> {});
        builder.show();
    }

    @Override
    public void injectFragment(AppComponent component) {
        component.inject(this);
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/fragments/social/QuestDetailFragment.kt
package com.habitrpg.android.habitica.ui.fragments.social

import android.annotation.SuppressLint
import android.app.AlertDialog
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.core.text.toHtml
import androidx.lifecycle.lifecycleScope
import com.habitrpg.android.habitica.R
import com.habitrpg.android.habitica.data.InventoryRepository
import com.habitrpg.android.habitica.data.SocialRepository
import com.habitrpg.android.habitica.databinding.FragmentQuestDetailBinding
import com.habitrpg.android.habitica.helpers.HapticFeedbackManager
import com.habitrpg.android.habitica.models.inventory.Quest
import com.habitrpg.android.habitica.models.inventory.QuestContent
import com.habitrpg.android.habitica.models.members.Member
import com.habitrpg.android.habitica.models.social.Group
import com.habitrpg.android.habitica.ui.fragments.BaseMainFragment
import com.habitrpg.android.habitica.ui.viewmodels.MainUserViewModel
import com.habitrpg.android.habitica.ui.views.dialogs.HabiticaAlertDialog
import com.habitrpg.common.habitica.extensions.fromHtml
import com.habitrpg.common.habitica.extensions.loadImage
import com.habitrpg.common.habitica.helpers.ExceptionHandler
import com.habitrpg.common.habitica.helpers.MarkdownParser
import com.habitrpg.common.habitica.helpers.launchCatching
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.filterNotNull
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch
import javax.inject.Inject

@AndroidEntryPoint
class QuestDetailFragment : BaseMainFragment<FragmentQuestDetailBinding>() {

    @Inject
    lateinit var socialRepository: SocialRepository

    @Inject
    lateinit var inventoryRepository: InventoryRepository

    @Inject
    lateinit var userViewModel: MainUserViewModel

    override var binding: FragmentQuestDetailBinding? = null

    override fun createBinding(
        inflater: LayoutInflater,
        container: ViewGroup?
    ): FragmentQuestDetailBinding {
        return FragmentQuestDetailBinding.inflate(inflater, container, false)
    }

    private var party: Group? = null
    private var quest: Quest? = null
    private var beginQuestMessage: String? = null

    private val isQuestActive: Boolean
        get() = quest?.active == true

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        this.hidesToolbar = true
        return super.onCreateView(inflater, container, savedInstanceState)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        showsBackButton = true
        super.onViewCreated(view, savedInstanceState)

        binding?.questBeginButton?.setOnClickListener { onQuestBegin() }
        binding?.questCancelButton?.setOnClickListener { onQuestCancel() }
        binding?.questLeaveButton?.setOnClickListener { onQuestLeave() }

        lifecycleScope.launch(ExceptionHandler.coroutine()) {
            userRepository.getUser()
                .map { it?.party?.id }
                .filterNotNull()
                .distinctUntilChanged()
                .flatMapLatest { socialRepository.getGroup(it) }
                .onEach { updateParty(it) }
                .map { it?.quest?.key }
                .filterNotNull()
                .distinctUntilChanged()
                .flatMapLatest { inventoryRepository.getQuestContent(it) }
                .filterNotNull()
                .collect {
                    updateQuestContent(it)
                }
        }
    }

    private fun updateParty(group: Group?) {
        if (binding?.titleView == null || group == null || group.quest == null) {
            return
        }
        party = group
        quest = group.quest
        setQuestParticipants(group.quest?.participants)
        lifecycleScope.launch(ExceptionHandler.coroutine()) {
            val member = socialRepository.retrieveMember(quest?.leader)
            if (context != null && binding?.questLeaderView != null) {
                binding?.questLeaderView?.text =
                    context?.getString(R.string.quest_leader_header, member?.displayName)
            }
        }

        val user = userViewModel.user.value
        if (binding?.questResponseWrapper != null) {
            if (userViewModel.userID != party?.quest?.leader && user?.party?.quest?.key == group.quest?.key && user?.party?.quest?.RSVPNeeded == false) {
                binding?.questLeaveButton?.visibility = View.VISIBLE
            } else {
                binding?.questLeaveButton?.visibility = View.GONE
            }
            if (showLeaderButtons()) {
                binding?.questCancelButton?.visibility = View.VISIBLE
                if (isQuestActive) {
                    binding?.questBeginButton?.visibility = View.GONE
                } else {
                    binding?.questBeginButton?.visibility = View.VISIBLE
                }
            } else {
                binding?.questCancelButton?.visibility = View.GONE
                binding?.questBeginButton?.visibility = View.GONE
            }
        }
    }

    private fun showLeaderButtons(): Boolean {
        return userViewModel.userID == party?.quest?.leader || userViewModel.userID == party?.leaderID
    }

    private fun updateQuestContent(questContent: QuestContent) {
        if (binding?.titleView == null || !questContent.isManaged) {
            return
        }
        binding?.titleView?.text = questContent.text
        // We need to do this, because the quest description can contain markdown AND HTML.
        binding?.descriptionView?.setText(
            MarkdownParser.parseMarkdown(questContent.notes).toHtml().fromHtml(),
            TextView.BufferType.SPANNABLE
        )

        binding?.questScrollImageView?.loadImage("inventory_quest_scroll_" + questContent.key)
    }

    private fun setQuestParticipants(participants: List<Member>?) {
        if (binding?.questParticipantList == null) {
            return
        }
        binding?.questParticipantList?.removeAllViews()
        val inflater = context?.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as? LayoutInflater
        var participantCount = 0
        for (participant in participants ?: emptyList()) {
            if (quest?.active == true && participant.participatesInQuest == false) {
                continue
            }
            val participantView =
                inflater?.inflate(R.layout.quest_participant, binding?.questParticipantList, false)
            val textView = participantView?.findViewById<View>(R.id.participant_name) as? TextView
            textView?.text = participant.displayName
            val statusTextView = participantView?.findViewById<View>(R.id.status_view) as? TextView
            if (quest?.active == false) {
                context?.let {
                    when (participant.participatesInQuest) {
                        null -> {
                            statusTextView?.setText(R.string.pending)
                            statusTextView?.setTextColor(
                                ContextCompat.getColor(
                                    it,
                                    R.color.text_ternary
                                )
                            )
                        }
                        true -> {
                            statusTextView?.setText(R.string.accepted)
                            statusTextView?.setTextColor(
                                ContextCompat.getColor(
                                    it,
                                    R.color.text_green
                                )
                            )
                        }
                        else -> {
                            statusTextView?.setText(R.string.declined)
                            statusTextView?.setTextColor(
                                ContextCompat.getColor(
                                    it,
                                    R.color.text_red
                                )
                            )
                        }
                    }
                }
                binding?.questParticipantList?.addView(participantView)
            } else {
                statusTextView?.visibility = View.GONE
                if (participant.participatesInQuest == true) {
                    binding?.questParticipantList?.addView(participantView)
                }
            }
            if (participant.participatesInQuest == true) {
                participantCount += 1
            }
        }
        if (quest?.active == true) {
            binding?.participantsHeader?.setText(R.string.participants)
            binding?.participantsHeaderCount?.text = participantCount.toString()
        } else {
            binding?.participantsHeader?.setText(R.string.invitations)
            @SuppressLint("SetTextI18n")
            binding?.participantsHeaderCount?.text =
                participantCount.toString() + "/" + participants?.size
            beginQuestMessage =
                getString(R.string.quest_begin_message, participantCount, participants?.size)
        }
    }

    override fun onDestroyView() {
        socialRepository.close()
        userRepository.close()
        inventoryRepository.close()
        super.onDestroyView()
    }

    private fun onQuestBegin() {
        HapticFeedbackManager.tap(requireView())
        val context = context
        if (context != null) {
            val alert = HabiticaAlertDialog(context)
            alert.setMessage(beginQuestMessage)
            alert.addButton(R.string.yes, true) { _, _ ->
                val party = party
                if (party != null) {
                    lifecycleScope.launchCatching {
                        socialRepository.forceStartQuest(party)
                    }
                }
            }
            alert.addButton(R.string.no, false)
            alert.show()
        }
    }

    private fun onQuestCancel() {
        HapticFeedbackManager.tap(requireView())
        context?.let {
            if (isQuestActive) {
                val builder = AlertDialog.Builder(activity)
                    .setMessage(R.string.quest_abort_message)
                    .setPositiveButton(R.string.yes) { _, _ ->
                        party?.id?.let { partyID ->
                            lifecycleScope.launchCatching {
                                socialRepository.abortQuest(partyID)
                                userRepository.retrieveUser(true)
                                activity?.supportFragmentManager?.popBackStack()
                            }
                        }
                    }.setNegativeButton(R.string.no) { _, _ -> }
                builder.show()
            } else {
                val alert = HabiticaAlertDialog(it)
                alert.setMessage(R.string.quest_cancel_message)
                alert.addButton(R.string.yes, true) { _, _ ->
                    party?.id?.let { partyID ->
                        lifecycleScope.launchCatching {
                            socialRepository.cancelQuest(partyID)
                            userRepository.retrieveUser(true)
                            activity?.supportFragmentManager?.popBackStack()
                        }
                    }
                }
                alert.addButton(R.string.no, false)
                alert.show()
            }
        }
    }

    private fun onQuestLeave() {
        HapticFeedbackManager.tap(requireView())
        val builder = AlertDialog.Builder(activity)
            .setMessage(if (quest?.active == true) R.string.quest_leave_message else R.string.quest_leave_message_nostart)
            .setPositiveButton(R.string.yes) { _, _ ->
                party?.id?.let { partyID ->
                    lifecycleScope.launchCatching {
                        socialRepository.leaveQuest(partyID)
                        socialRepository.retrieveGroup(partyID)
                        userRepository.retrieveUser(true)
                        activity?.supportFragmentManager?.popBackStack()
                    }
                }
            }.setNegativeButton(R.string.no) { _, _ -> }
        builder.show()
    }
}
